### PR TITLE
Expose getCurrentBranch() method on git connection + rework gitvalueProviderFactory

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?

It brings new method to update gitValueProviderFactory 
it will avoid to use status object requiring a lot of I/O
### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/2567
### New behavior

Faster way to grab current branch name
### PR type
- [x] Minor change = no change to existing features or docs
### Minor change checklist
- [ ] New API required?
- [ ] API updated
- [x] Tests provided / updated
- [x] Tests passed

Change-Id: If88da1e1b63a252f0226f1a586f3a67d7295e6c3
Signed-off-by: Florent BENOIT fbenoit@codenvy.com

### CODE REVIEW

File: wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java [hunk 1]
```diff
@@ -484,4 +484,14 @@
...
+
+
+    /**
+     * Get the current branch on the current directory
+     *
+     * @return the name of the branch
+     * @throws GitException
+     *         if any exception occurs
+     */
+    String getBranchName() throws GitException;
```
Comment (sleshchenko) [inline wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java]: IMHO `getBranchName` looks unclear enough. I think I'd much better to rename it to `getCurrentBranchName` or `getCurrentBranch` WDYT?

File: wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java [hunk 2]
```diff
@@ -484,4 +484,14 @@
 
     /** Set publisher for git output, e.g. for sending git command output to the client side. */
     void setOutputLineConsumerFactory(LineConsumerFactory outputPublisherFactory);
+
+
```
Comment (sleshchenko) [inline wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java]: please remove extra empty line

**Review by skabashnyuk — COMMENTED** [review]
ok for me. @vinokurig @sleshchenko  please approve.

**Review by sleshchenko — APPROVED** [review]

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java [hunk 3]
```diff
@@ -145,4 +148,20 @@ public void shouldDoNothingWhenTransportHttpReceived() throws Exception{
...
     }
+
+    /**
+     * Check branch using current repository reference is returned
+     * @throws Exception if it fails
+     */
+    @Test
+    public void checkBranchName() throws Exception {
+        String branchTest = "helloWorld";
+        Ref ref = Mockito.mock(Ref.class);
```
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java]: Please add static import to `Mockito.mock`

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java [hunk 4]
```diff
@@ -145,4 +148,20 @@ public void shouldDoNothingWhenTransportHttpReceived() throws Exception{
         //then
         verifyZeroInteractions(transportHttp);
     }
+
+    /**
+     * Check branch using current repository reference is returned
+     * @throws Exception if it fails
+     */
+    @Test
+    public void checkBranchName() throws Exception {
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java]: Please divide  the test to given, when then blocks, like in tests above

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java (hunk 3 above)
Comment (benoitf) [reply wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java]: ok

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (benoitf) [reply wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java]: Hum method name is subjective. It actually checks something so I can use check in it

Also it's divided in given/then I don't see any improvement on adding //given on mock() call

File: wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java (hunk 2 above)
Comment (benoitf) [reply wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java]: updated

File: wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java (hunk 1 above)
Comment (benoitf) [reply wsagent/che-core-api-git/src/main/java/org/eclipse/che/api/git/GitConnection.java]: updated

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java (hunk 3 above)
Comment (benoitf) [reply wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java]: updated

**Review by vinokurig — APPROVED** [review]

**Pull request merged**